File: cpp/069.Sqrt/Sqrt.cpp
/*
解题思路：
    两分查找
    类似于二分查找的变体：查找最后一个小于等于给定值的元素

*/

// O(logn)
// Runtime Beats 100%

#include <cassert>
#include <iostream>

class Solution {
public:
    int mySqrt(int x) {
        if (x < 2) {
            return x;
        }

        int left = 1;
        int right = x;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (x / mid < mid) {
                right = mid - 1;
            } else {
                if (x / mid >= mid && x / (mid + 1) < mid + 1) {
                    return mid;
                }

                left = mid + 1;
            }
        }
        return -1;
    }
};

void test1() {
    int x = 4;
    Solution s;
    assert(s.mySqrt(x) == 2);
}

void test2() {
    int x = 8;
    Solution s;
    assert(s.mySqrt(x) == 2);
}

int main() {
    test1();
    test2();
    return 0;
}

File: go/069.Sqrt/Sqrt.go
/*
解题思路：
    二分查找
    类似于二分查找的变体：查找最后一个小于等于给定值的元素
*/

// O(log n)
// Runtime Beats 100%

func mySqrt(x int) int {
	if x == 0 || x == 1 {
		return x
	}
	low := 1
	high := x
	res := 0
	for low <= high {
		mid := low + (high-low)>>1
		if mid == x/mid {
			return mid
		} else if mid < x/mid {
			low = mid + 1
			res = mid
		} else {
			high = mid - 1
		}
	}
	return res
}